File: 2019/day07.py
from itertools import permutations
from queue import Queue

class operation:
    def __init__(self, val):
        self.opcode = val % 100
        val //= 100
        self.immediate = []
        for x in range(3):
            self.immediate.append(val % 10 == 1)
            val //= 10

class Program:
    def __init__(self, program, phase):
        self.program = program.copy()
        self.pos = 0
        self.input = Queue()
        self.input.put(phase)

    def add_input(self, n):
        self.input.put(n)

    def finished(self):
        return self.program[self.pos] == 99

    def __next__(self):
        read = 0

        while not self.finished():
            op = operation(self.program[self.pos])
            
            if op.opcode == 1:
                # addition
                a = self.program[self.pos+1] if op.immediate[0] else self.program[self.program[self.pos+1]]
                b = self.program[self.pos+2] if op.immediate[1] else self.program[self.program[self.pos+2]]
                target = self.pos+3 if op.immediate[2] else self.program[self.pos+3]

                self.program[target] = a+b;
                self.pos+= 4
            elif op.opcode == 2:
                # multiplication
                a = self.program[self.pos+1] if op.immediate[0] else self.program[self.program[self.pos+1]]
                b = self.program[self.pos+2] if op.immediate[1] else self.program[self.program[self.pos+2]]
                target = self.pos+3 if op.immediate[2] else self.program[self.pos+3]

                self.program[target] = a*b;
                self.pos+= 4
            elif op.opcode == 3:
                # input
                target = self.pos+1 if op.immediate[0] else self.program[self.pos+1]
                self.program[target] = self.input.get()
                self.pos += 2
            elif op.opcode == 4:
                # output
                target = self.pos+1 if op.immediate[0] else self.program[self.pos+1]
                self.pos += 2
                return self.program[target]
            elif op.opcode == 5:
                # jump if true
                a = self.program[self.pos+1] if op.immediate[0] else self.program[self.program[self.pos+1]]
                if a == 0:
                    self.pos += 3
                else:
                    self.pos = self.program[self.pos+2] if op.immediate[1] else self.program[self.program[self.pos+2]]
            elif op.opcode == 6:
                # jump if false
                a = self.program[self.pos+1] if op.immediate[0] else self.program[self.program[self.pos+1]]
                if a != 0:
                    self.pos += 3
                else:
                    self.pos = self.program[self.pos+2] if op.immediate[1] else self.program[self.program[self.pos+2]]
            elif op.opcode == 7:
                # less than
                a = self.program[self.pos+1] if op.immediate[0] else self.program[self.program[self.pos+1]]
                b = self.program[self.pos+2] if op.immediate[1] else self.program[self.program[self.pos+2]]
                target = self.pos+3 if op.immediate[2] else self.program[self.pos+3]

                self.program[target] = 1 if a < b else 0
                self.pos+= 4
            elif op.opcode == 8:
                # equals
                a = self.program[self.pos+1] if op.immediate[0] else self.program[self.program[self.pos+1]]
                b = self.program[self.pos+2] if op.immediate[1] else self.program[self.program[self.pos+2]]
                target = self.pos+3 if op.immediate[2] else self.program[self.pos+3]

                self.program[target] = 1 if a == b else 0
                self.pos+= 4
            else:
                # error
                print('Invalid opcode!')
                exit(1)
        raise StopIteration()

def test():
    # day 5 test
    program = [int(n) for n in input().split(',')]
    data = compute(None, 5, program)
    print('ans:', data)

def part_one():
    program = [int(n) for n in input().split(',')]
    phase_base = [0,1,2,3,4]
    maximum = -10**30
    for phases in permutations(phase_base):
        data = 0
        for i in range(5):
            data = compute(data, phases[i], program)
        maximum = max(maximum, data)
    print('ans:', maximum)

def part_two():
    program = [int(n) for n in input().split(',')]
    phase_base = [5,6,7,8,9]
    maximum = -10**30
    for phases in permutations(phase_base):
        amps = [Program(program, phases[x]) for x in range(5)]
        amps[0].add_input(0)
        i = 0
        signal = 0
        while True:
            try:
                output = next(amps[i])
            except StopIteration:
                break
            if i == 4:
                signal = output
                amps[0].add_input(output)
                i = 0
            else:
                i += 1
                amps[i].add_input(output)
        maximum = max(maximum, signal)
    print('answer:', maximum)

        
if __name__ == '__main__':
    part_two()

File: 2019/day07.cc
#include <iostream>
#include <vector>
#include <string>
#include <cassert>


struct operation {
	int opcode;
	bool immediate[3];

	operation(int val) {
		opcode = val % 100;
		val /= 100;
		for (int i = 0; i != 3; i++) {
			immediate[i] = val % 10 == 1;
			val /= 10;
		}
	}
};

void compute(int input, int phase) {
	size_t sz = 0;
	int program[700];

	for (int i; std::cin >> i; ) {
		program[sz] = i;
		sz++;
		if (std::cin.peek() == ',') {
			std::cin.ignore();
		}
	}

	//for (int i = 0; i != sz; ++i) {
		//std::cout << program[i] << ",";
	//}
	
	int input;
	std::cin >> input;
	std::cout << "Input: " << input << "\n";
	program[sz] = 0;
	bool read = false;
	int n = 0;

	int pos = 0;
	while(program[pos] != 99) {
		operation op = *(new operation(program[pos]));
		
		if (op.opcode == 1) {
			int a = op.immediate[0] ? program[pos+1] : program[program[pos+1]];
			int b = op.immediate[1] ? program[pos+2] : program[program[pos+2]];
			int* target = op.immediate[2] ? &program[pos+3] : &program[program[pos+3]];

			target[0] = a + b;
			pos += 4;
		} else if (op.opcode == 2) {
			 //multiplication
			int a = op.immediate[0] ? program[pos+1] : program[program[pos+1]];
			int b = op.immediate[1] ? program[pos+2] : program[program[pos+2]];
			int* target = op.immediate[2] ? &program[pos+3] : &program[program[pos+3]];

			target[0] = a * b;
			pos += 4;
		} else if (op.opcode == 3) {
			 //input
			 assert(!read);
			 int* target = op.immediate[0] ? &program[pos+1] : &program[program[pos+1]];
			 target[0] = input;
			 read = true;
			 pos += 2;
		} else if (op.opcode == 4) {
			 //output
			int* target = op.immediate[0] ? &program[pos+1] : &program[program[pos+1]];
			std::cout << target[0] << '\n';
			pos += 2;
		} else if (op.opcode == 5) {
			// jump if true
			int a = op.immediate[0] ? program[pos+1] : program[program[pos+1]];
			if (a != 0) {
				pos = op.immediate[1] ? program[pos+2] : program[program[pos+2]];
			} else {
				pos += 3;
			}
		} else if (op.opcode == 6) {
			// jump if false
			int a = op.immediate[0] ? program[pos+1] : program[program[pos+1]];
			if (a == 0) {
				pos = op.immediate[1] ? program[pos+2] : program[program[pos+2]];
			} else {
				pos += 3;
			}
		} else if (op.opcode == 7) {
			// less than
			int a = op.immediate[0] ? program[pos+1] : program[program[pos+1]];
			int b = op.immediate[1] ? program[pos+2] : program[program[pos+2]];
			int* target = op.immediate[2] ? &program[pos+3] : &program[program[pos+3]];

			target[0] = a < b ? 1 : 0;
			pos += 4;
		} else if (op.opcode == 8) {
			// equals
			int a = op.immediate[0] ? program[pos+1] : program[program[pos+1]];
			int b = op.immediate[1] ? program[pos+2] : program[program[pos+2]];
			int* target = op.immediate[2] ? &program[pos+3] : &program[program[pos+3]];

			target[0] = a == b ? 1 : 0;
			pos += 4;
		} else {
			std::cout << "invalid opcode at pos " << pos << '\n';
			std::cout << "opcode: " << op.opcode << " Immediates: " << op.immediate[0] << ", " << op.immediate[1] << ", " << op.immediate[2] << '\n';
			exit(1);
		}
		++n;
		//std::cout << pos << '\n';
	}
}

int main() {
	part2();
}
